clusterloader2/pkg/measurement/common/network/network_performance_measurement.go

```diff
@@ -0,0 +1,506 @@
+	stopCh := make(chan struct{})
+	time.AfterFunc(weightedPodReadyTimeout, func() {
+		close(stopCh)
+	})
+	options := &measurementutil.WaitForPodOptions{
+		Selector:            &measurementutil.ObjectSelector{Namespace: netperfNamespace},
+		DesiredPodCount:     npm.numberOfClients + npm.numberOfServers,
+		CallerName:          networkPerformanceMetricsName,
+		WaitForPodsInterval: podReadyPollInterval,
+	}
+	if err := measurementutil.WaitForPods(npm.k8sClient, stopCh, options); err != nil {
+		return err
+	}
+	return nil
+}
+
+// getWeightedTimerValuesForPoll returns poll timeout value.
+// Pod numbers are taken into consideration to decide the timer values.
+func (npm *networkPerformanceMeasurement) getWeightedPodReadyTimeout() time.Duration {
+	weightedPodReadyTimeout := float32(npm.numberOfServers+npm.numberOfClients) * podReadyTimeOutFactor
+	podReadyTimeout := time.Duration(weightedPodReadyTimeout) * time.Second
+
+	klog.Info("podReadyTimeout: ", podReadyTimeout)
+	return podReadyTimeout
+}
+
+func (*networkPerformanceMeasurement) String() string {
+	return networkPerformanceMetricsName
+}
+
+func (npm *networkPerformanceMeasurement) storeWorkerPods() error {
+	var podsWithIPs int
+	options := metav1.ListOptions{}
+	pods, err := npm.k8sClient.CoreV1().Pods(netperfNamespace).List(context.TODO(), options)
+	if err != nil {
+		return err
+	}
+	for _, pod := range pods.Items {
+		if pod.Status.PodIP != "" {
+			podData := &workerPodData{podName: pod.Name, podIP: pod.Status.PodIP, workerNode: pod.Spec.NodeName}
+			npm.populateWorkerPodList(podData)
+			podsWithIPs++
+		}
+	}
+	klog.Infof("Pods configured: %v, Pods with IPs: %v", npm.numberOfServers+npm.numberOfClients, podsWithIPs)
+	return nil
+}
+
+func (npm *networkPerformanceMeasurement) gather() (measurement.Summary, error) {
+	npm.createResultSummary()
+	content, err := util.PrettyPrintJSON(&measurementutil.PerfData{
+		Version:   "v1",
+		DataItems: npm.resultSummary.DataItems,
+	})
+	if err != nil {
+		klog.Info("Pretty Print to Json Err: ", err)
+	}
+	summaryName := fmt.Sprintf(npm.String() + npm.resultSummary.podRatio + npm.resultSummary.protocol + npm.resultSummary.service)
+	return measurement.CreateSummary(summaryName, "json", content), nil
+}
+
+// Dispose disposes resources,objects after the measurement.
+func (npm *networkPerformanceMeasurement) Dispose() {
+	if npm.framework == nil {
+		klog.V(1).Infof("Network measurement %s wasn't started, skipping the Dispose() step", npm)
+		return
+	}
+	klog.Infof("Stopping network performance measurement...")
+	close(npm.stopCh)
+	close(npm.podInfo.podPairCh)
+	npm.cleanupCluster()
+}
+
+func (npm *networkPerformanceMeasurement) handleUpdateNetworkTestEvents(newObj interface{}) {
+	newRuntimeObj, ok := newObj.(runtime.Object)
+	if newObj != nil && !ok {
+		klog.Errorf("error casting object: %s", newObj)
+		return
+	}
+
+	resourceContent, err := runtime.DefaultUnstructuredConverter.ToUnstructured(newRuntimeObj)
```

Comment:
Instead of operating on Unstructured items, you basically want to define your type the same way built-in APIs are defined, (take a look at the example):
https://github.com/kubernetes/sample-controller/blob/master/pkg/apis/samplecontroller/v1alpha1/types.go

[you may want to autogenerate the remaining stuff there too]

Reply:
I'm not asking to crate any controller or operator.

I'm asking you to define the type. It's not defined now. Basically you do types.go and pretty much everything else is autogenerated. Operating on unstructured (if not needed), isn't a good pattern - thi  code would be much simpler if operating on the type itself.

Reply:
In measurement , i am only reading the sub-object "status" unlike worker pod where he needs all the info present in NetworkTestRequest CR. We can define a type and autogenerate. Then i think both measurement and worker needs to modify the code again. Do we need to do in this PR or follow up PR ?

Reply:
Yeah - I don't want to block this PR indefinitely.
Please open an issue to do that migration and add here:
TODO(#<issue number>): Stop relying on unstructured data.